src/main/java/org/opensearch/index/codec/customcodecs/Lucene99QatCodec.java

```diff
@@ -0,0 +1,147 @@
+     *
+     * @param mode The compression codec (QAT_DEFLATE or QAT_LZ4).
+     */
+    public Lucene99QatCodec(Mode mode) {
+        this(mode, DEFAULT_COMPRESSION_LEVEL);
+    }
+
+    /**
+     * Creates a new compression codec with the given compression level. We use lowercase letters when
+     * registering the codec so that we remain consistent with the other compression codecs: default,
+     * lucene_default, and best_compression.
+     *
+     * @param mode The compression codec (QAT_DEFLATE or QAT_LZ4).
+     * @param compressionLevel The compression level.
+     */
+    public Lucene99QatCodec(Mode mode, int compressionLevel) {
+        super(mode.getCodec(), new Lucene99Codec());
+        this.storedFieldsFormat = new Lucene99QatStoredFieldsFormat(mode, compressionLevel);
+    }
+
+    /**
+     * Creates a new compression codec with the given compression level. We use lowercase letters when
+     * registering the codec so that we remain consistent with the other compression codecs: default,
+     * lucene_default, and best_compression.
+     *
+     * @param mode The compression codec (QAT_DEFLATE, or QAT_LZ4).
+     * @param compressionLevel The compression level.
+     * @param supplier supplier for QAT mode.
+     */
+    public Lucene99QatCodec(Mode mode, int compressionLevel, Supplier<QatZipper.Mode> supplier) {
+        super(mode.getCodec(), new Lucene99Codec());
+        this.storedFieldsFormat = new Lucene99QatStoredFieldsFormat(mode, compressionLevel, supplier);
+    }
+
+    /**
+     * Creates a new compression codec with the given compression level. We use lowercase letters when
+     * registering the codec so that we remain consistent with the other compression codecs: default,
+     * lucene_default, and best_compression.
+     *
+     * @param mode The compression codec (QAT_DEFLATE or QAT_LZ4).
+     * @param compressionLevel The compression level.
+     * @param mapperService The mapper service.
+     * @param logger The logger.
+     */
+    public Lucene99QatCodec(Mode mode, int compressionLevel, MapperService mapperService, Logger logger) {
+        super(mode.getCodec(), new PerFieldMappingPostingFormatCodec(Lucene99Codec.Mode.BEST_SPEED, mapperService, logger));
+        this.storedFieldsFormat = new Lucene99QatStoredFieldsFormat(mode, compressionLevel);
+    }
+
+    /**
+     * Creates a new compression codec with the given compression level. We use lowercase letters when
+     * registering the codec so that we remain consistent with the other compression codecs: default,
+     * lucene_default, and best_compression.
+     *
+     * @param mode The compression codec (ZSTD, ZSTD_NO_DICT, QAT_DEFLATE, or QAT_LZ4).
+     * @param compressionLevel The compression level.
+     * @param mapperService The mapper service.
+     * @param logger The logger.
+     * @param supplier supplier for QAT mode.
+     */
+    public Lucene99QatCodec(
+        Mode mode,
+        int compressionLevel,
```

Comment:
@mulerm @mulugetam Can you please elaborate on how the compression level works here. Do the algorithm provide that abstraction to configure compression level?

Reply:
The compression level is ultimately passed into QatZipper which does the compression.

Reply:
Got it, but trying to understand how the levels affect the compression for both of the algorithms? What would be the trade offs? What would be the optimal settings? How would it affect the store size? 
I was not aware that we can control compression ratio for LZ4 and Deflate algorithms as well.

Reply:
Do we also need to restrict the range of compression level?

Reply:
We have a restriction on the compression level [over here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L171) in OpenSearch.
This was kept keeping Zstandard in mind. 

If the range in QAT is not kept inline (the compression ratio expected with each level) with Zstandard, then it would be confusing behaviour for the users.

Tagging: @backslasht @reta @mgodwan for visibility

Reply:
> Tagging: @backslasht @reta @mgodwan for visibility

Thanks @sarthakaggarwal97 , I think we could fix that: `CodecSettings` could be extended to validate the setting value per codec (right now it is only validates that setting is support but not its value).

Reply:
@reta do you mean that we would not need to change the value of the min - max levels in the Setting.IntegerParser?

Reply:
So here is a brief idea: ideally, compression levels should be specific to codes. Just for illustration purposes, assume ZSTD has 1 to 6, and LZ4 1 to 10. The INDEX_CODEC_COMPRESSION_LEVEL_SETTING has right now range from 1 to 6 hardcoded (since we didn't thought about that). However, we could move off the range min/max validation to each codec instead.

Reply:
@mulerm are we okay with current compression level range of (1,6) or do we feel the need to extend it?

Reply:
> @mulerm are we okay with current compression level range of (1,6) or do we feel the need to extend it?

We're okay with this for now. Going forward, compression levels should be specific to codecs, as @reta suggested.